Find_first_missing_positive_integer.rb

```diff
@@ -0,0 +1,27 @@
+=begin 
+This problem was asked by Stripe.
+
+Given an array of integers, find the first missing positive integer in linear time and constant space. 
+In other words, find the lowest positive integer that does not exist in the array. 
+The array can contain duplicates and negative numbers as well.
+
+For example, the input [3, 4, -1, 1] should give 2. The input [1, 2, 0] should give 3.
+
+You can modify the input array in-place.
+=end 
+
+def missingpostiveint(array)
+    array = array.sort 
+    array.each_with_index do |value, index|
+    return (value + 1) if value >= 1 && array[index + 1] != value + 1  
+    return (value + 1) if value <-1  && array [index + 1] != value + 1
+    end

```

Comment:
You can simplify this. 

Try something like this:

1. Loop through the array. 
2. Each time calculate what the next number should be. If the next number is now what you expect, then return that number that is missing. 

```
expected = array[0]
array.each do |value|
    return expected if value != expected
    expected += 1
end
```

Reply:
I think you are still thinking too complex. Think exactly like how your brain will think naturally... whatever we would do naturally is usually the easiest algorithm and you can just easily code it that way also. 

Like if you put a set of numbers before me, I am thinking about what should be the next number and checking if that next number is what I am looking for. That how our brains naturally work. You just have to slow down the brain process to then come up with the simplest algorithm.

Reply:
To be honest that's true. Before i jump into the logic or the algorithm, i tend mostly  to jump into the code which is not good. But I'm gradually changing from that habit now.

Reply:
> Try something like this:
> 
> 1. Loop through the array.
> 2. Each time calculate what the next number should be. If the next number is now what you expect, then return that number that is missing.
> 
> ```
> expected = array[0]
> array.each do |value|
>     return expected if value != expected
>     expected += 1
> end
> ```

This is working all the arrays except for this array [3, 4, -1, 1] it returns 5 instead of 2.